script/src/scheduler.rs

```diff
@@ -340,6 +339,29 @@ where
         let messages: Vec<Message> = self.message_box.lock().expect("lock").drain(..).collect();
         for message in messages {
             match message {
+                Message::ExecV2(vm_id, args) => {
+                    let (_, old_machine) = self
+                        .instantiated
+                        .get_mut(&vm_id)
+                        .ok_or_else(|| Error::Unexpected("Unable to find VM Id".to_string()))?;
+                    old_machine
+                        .machine
+                        .add_cycles(EXEC_LOAD_ELF_V2_CYCLES_BASE)?;
+                    let old_cycles = old_machine.machine.cycles();
+                    let max_cycles = old_machine.machine.max_cycles();
+                    let (context, mut new_machine) = self.create_dummy_vm(&vm_id)?;
+                    new_machine.set_max_cycles(max_cycles);
+                    new_machine.machine.add_cycles_no_checking(old_cycles)?;
+                    self.load_vm_program(
+                        &context,
+                        &mut new_machine,
+                        &args.data_piece_id,
+                        args.offset,
+                        args.length,
+                        Some((vm_id, args.argc, args.argv)),
+                    )?;
+                    self.instantiated.insert(vm_id, (context, new_machine));
```

Comment:
Please cap the number of maximum instantiated virtual machines here.

Reply:
Exec will always remove an old running vm instance and then join a new one, so the maximum number of instantiations will not be triggered.

Reply:
Ah I see, in that sense this code is correct, but can we change the code to the following flow to better illustrate this fact:

```
let machine_n_context = self.instantiated.get_mut(&vm_id)...
{
  let old_machine = &machine_n_context.1;
  ...
}
...
*machine_n_context = (context, new_machine);
```

The latter insert line can be quite confusing and maybe one day we forgot about this underlying logic.

Reply:
In this way of writing, `machine_n_context` borrows `self` as mutable, This will cause `self.create_dummy_vm` and `self.load_vm_program` to no longer be usable

Reply:
I see, you are right here. In this sence I recommend the following line as a hint:

```
debug_assert!(self.instantiated.contains_key(&vm_id));
self.instantiated.insert(vm_id, (context, new_machine));
```

Reply:
OK. I'll add a comment as well.